### ISSUE TITLE
Crash: Attempt to invoke virtual method 'void java.net.DatagramSocket.setReuseAddress(boolean)' on a null object reference

### ISSUE THREAD

**Issue by hubciorz** [tradle/react-native-udp#147]
## Description
App crashes if creating socket object is not successful (object equals to null). This issue is basically a copy of #119 as @Rapsssito asked to create a new one.

## Error
```java
com.tradle.react.UdpSocketClient.bind
UdpSocketClient.java, line 83
java.lang.NullPointerException: Attempt to invoke virtual method 'void java.net.DatagramSocket.setReuseAddress(boolean)' on a null object reference
```

## Stack traces
```java
com.tradle.react.UdpSocketClient.bind UdpSocketClient.java:83
com.tradle.react.UdpSockets$2.run UdpSockets.java:118
java.lang.Thread.run Thread.java:919
java.util.concurrent.ThreadPoolExecutor.runWorker ThreadPoolExecutor.java:1167
java.util.concurrent.ThreadPoolExecutor$Worker.run ThreadPoolExecutor.java:641
java.lang.Thread.run Thread.java:919
```

## Steps to reproduce
Probably, you need to create and close sockets on and on. It is really hard to reproduce quickly. 

## Current behavior
App crashes on some devices with multiple Android versions. Sometimes the error is `Attempt to invoke virtual method 'void java.net.DatagramSocket.send(java.net.DatagramPacket)' on a null object reference` but the reason seems to be the same.

## Expected behavior
App does not crash. The solution proposed by @rimzici is to surround the code responsible for crashing with a try/catch block (see the code in his fork: https://github.com/telldus/react-native-udp/blob/master/android/src/main/java/com/tradle/react/UdpSocketClient.java#L83-L91). However, maybe a better idea would be a simple null-checking? 

## Relevant information
|             |     |
| ----------- | --- |
| react-native | 0.63.4   |
| react-native-udp  | 4.0.4 (but happens on earlier versions, too)   |

## Occurred on
|             |     |
| ----------- | --- |
| Device | OS   |
| Galaxy S7  | Android 8   |
| Galaxy S9  | Android 10   |
| Galaxy S9+  | Android 10  |
| Galaxy S10+  | Android 10  |

**Comment by Rapsssito** [comment]
@hubciorz, could you check if the issue persist on the branch named `fix-147`?
```
yarn add https://github.com/tradle/react-native-udp#fix-147
```

**Comment by hubciorz** [comment]
@Rapsssito Sure. We are going to release an update to our app on Monday, so we will need then a couple of days to see if there are any crash reports. Thanks!

**Comment by Incrementing** [comment]
I had this crash last night on a Samsung Galaxy S8+. Now using branch fix-147, I'll also keep an eye out for this occurring again and post an update 👍

**Comment by hubciorz** [comment]
@Rapsssito 
We did not notice the original error described on the first comment but now we have a different error located in the same place (so the fix introduced a regression):

```java
com.tradle.react.UdpSocketClient.bind
UdpSocketClient.java, line 84
java.lang.IllegalStateException: Cannot execute task: the task has already been executed (a task can be executed only once)
```

We still can see "variants" of the same kind of error (I mentioned about them later on my first comment):

```java
com.tradle.react.UdpReceiverTask.doInBackground
UdpReceiverTask.java, line 45
java.lang.NullPointerException: Attempt to invoke virtual method 'void java.net.DatagramSocket.receive(java.net.DatagramPacket)' on a null object reference
```

I know that try/catch does not solve the roots of the issue but at least it protects the app from crashing.

**Comment by Rapsssito** [comment]
@hubciorz feel free to add the try/catch fix using any package patcher (such as `patch-package`). However, that won't fix the real issue and won't fix `NullPointerException` if you want to use the socket.

**Comment by Rapsssito** [comment]
@hubciorz, I am working on a new fix.

**Comment by Rapsssito** [comment]
@hubciorz I have commited new changes to the branch, could check if it works that way? It the issue persist I might just create a null check and throw an exception.

**Comment by hubciorz** [comment]
@Rapsssito We do not plan to release any app update before Christmas or New Year's Eve so it has to wait until the first week of January.

**Comment by hubciorz** [comment]
@Rapsssito I hope you had a nice Christmas!

We released an app update a couple of days ago and it seems to be much better. Just one crash in the new implementation so far:

## Error
```java
libcore.io.Posix.getsockoptLinger
Posix.java, line -2
java.util.concurrent.TimeoutException: java.net.PlainSocketImpl.finalize() timed out after 10 seconds
```

## Stack traces
```java
libcore.io.Posix.getsockoptLinger Posix.java
libcore.io.ForwardingOs.getsockoptLinger ForwardingOs.java:94
libcore.io.BlockGuardOs.isLingerSocket BlockGuardOs.java:105
libcore.io.BlockGuardOs.close BlockGuardOs.java:87
libcore.io.IoBridge.closeAndSignalBlockedThreads IoBridge.java:203
java.net.PlainSocketImpl.close PlainSocketImpl.java:148
java.net.PlainSocketImpl.finalize PlainSocketImpl.java:206
java.lang.Daemons$FinalizerDaemon.doFinalize Daemons.java:202
java.lang.Daemons$FinalizerDaemon.run Daemons.java:185
java.lang.Thread.run Thread.java:818
```

## Occurred on
|             |     |
| ----------- | --- |
| Device | OS   |
| Nexus 7  | Android 6.0.1   |

---

Tested by about 2000 users. I will keep you updated if I have more crash reports.

**Comment by Rapsssito** [comment]
@hubciorz, **glad to hear that!** I have been reading about the issue and it looks like it might be non-fixable as it has to do with the kernel context switching. Others have experience the same type of issue using other kind of libraries: [stackoverflow thread](https://stackoverflow.com/questions/24021609/how-to-handle-java-util-concurrent-timeoutexception-android-os-binderproxy-fin).

If the issue persists in other devices, I will take a closer look. However, as it is just one crash, I would consider this issue's initial problem fixed. Feel free to close this issue when you feel confident enough that it is not happening again.

**Comment by hubciorz** [comment]
@Rapsssito Thank you for your hard work. It seems that #144 has also been fixed. 

Let give the fix a couple of days more. We currently have about 18 000 active users so so far about ~15% of them installed and run the app. As I said, I will inform you about any crashes.

Thanks!

**Comment by hubciorz** [comment]
@Rapsssito Please re-open :( Unfortunately, the problem is still there (same for #144). I am not sure if the issues occurs less often or if it is a placebo effect, but I noticed a couple of crashes last weekend.

**Comment by Rapsssito** [comment]
@hubciorz, this issue might be too long for people experiencing this same problem. If you are receiving the same new crash report, could you please open a new issue with the updated information?

**Comment by haveneersrobin** [comment]
We see a similar issue: `NullPointerException: Attempt to invoke virtual method 'void java.net.DatagramSocket.send(java.net.DatagramPacket)' on a null object referencez` with version `4.1.3`

cc @Rapsssito : do I open a new issue for this?

**Comment by Rapsssito** [comment]
@haveneersrobin, yes, please. Open a new issue.
